File: shared/envelope/replay.go
package envelope

import (
	"container/list"
	"sync"
	"time"
)

// replayEntry holds a jobID and the time it was observed.
type replayEntry struct {
	jobID      string
	observedAt time.Time
}

// ReplayCache is a 24-hour, fixed-size, goroutine-safe ring buffer of seen
// jobIds. It rejects duplicates (architecture §3.4 — replay protection).
//
// Implementation: a hash map for O(1) lookup + a doubly-linked list for
// ordered eviction (oldest-first). RSS target: ≤ 10 MB for 8.64M entries.
type ReplayCache struct {
	mu       sync.RWMutex
	capacity int
	ttl      time.Duration
	index    map[string]*list.Element // jobID → list element
	order    *list.List               // front = oldest
	stopCh   chan struct{}
}

// NewReplayCache creates a replay cache with the given capacity and TTL.
// A background goroutine purges expired entries every minute.
// Call Close() when done (e.g. in tests) to stop the goroutine.
func NewReplayCache(capacity int, ttl time.Duration) *ReplayCache {
	c := &ReplayCache{
		capacity: capacity,
		ttl:      ttl,
		index:    make(map[string]*list.Element, capacity),
		order:    list.New(),
		stopCh:   make(chan struct{}),
	}
	go c.purgeLoop()
	return c
}

// Seen returns true if jobID has been seen within the TTL window.
func (c *ReplayCache) Seen(jobID string) bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	el, ok := c.index[jobID]
	if !ok {
		return false
	}
	entry := el.Value.(*replayEntry)
	return time.Since(entry.observedAt) < c.ttl
}

// Remember records jobID as seen. If the cache is at capacity, the oldest
// entry is evicted to make room (oldest = lowest insertion order, regardless
// of TTL). This matches the spec: evict oldest, not newest.
func (c *ReplayCache) Remember(jobID string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Already in cache — update timestamp and move to back (most recent).
	if el, ok := c.index[jobID]; ok {
		el.Value.(*replayEntry).observedAt = time.Now()
		c.order.MoveToBack(el)
		return
	}

	// Evict oldest if at capacity.
	if c.order.Len() >= c.capacity {
		oldest := c.order.Front()
		if oldest != nil {
			entry := oldest.Value.(*replayEntry)
			delete(c.index, entry.jobID)
			c.order.Remove(oldest)
		}
	}

	// Insert new entry at the back (most recent).
	el := c.order.PushBack(&replayEntry{
		jobID:      jobID,
		observedAt: time.Now(),
	})
	c.index[jobID] = el
}

// Close stops the background purge goroutine. Safe to call multiple times.
func (c *ReplayCache) Close() {
	select {
	case <-c.stopCh:
		// already closed
	default:
		close(c.stopCh)
	}
}

// Len returns the current number of entries (for testing).
func (c *ReplayCache) Len() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.order.Len()
}

// purgeLoop runs in the background and removes entries older than TTL every
// minute. It walks the list from front (oldest) until it hits a non-expired
// entry, then stops — no need to traverse further because entries are in
// insertion order.
func (c *ReplayCache) purgeLoop() {
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()
	for {
		select {
		case <-c.stopCh:
			return
		case <-ticker.C:
			c.purge()
		}
	}
}

func (c *ReplayCache) purge() {
	c.mu.Lock()
	defer c.mu.Unlock()
	cutoff := time.Now().Add(-c.ttl)
	for {
		front := c.order.Front()
		if front == nil {
			break
		}
		entry := front.Value.(*replayEntry)
		if entry.observedAt.After(cutoff) {
			break // rest of list is newer
		}
		delete(c.index, entry.jobID)
		c.order.Remove(front)
	}
}

File: shared-rust/src/replay.rs
//! Replay-protection ring buffer — mirrors agents/shared/envelope/replay.go.

use indexmap::IndexMap;
use parking_lot::RwLock;
use std::sync::Arc;
use std::time::{Duration, Instant};

/// A goroutine-safe, fixed-capacity ring buffer of seen job IDs.
///
/// Uses `IndexMap` for insertion-ordered O(1) eviction of the oldest entry
/// and `parking_lot::RwLock` for low-contention reads.
#[derive(Clone)]
pub struct ReplayCache {
    inner: Arc<RwLock<Inner>>,
}

struct Inner {
    capacity: usize,
    ttl: Duration,
    map: IndexMap<String, Instant>, // job_id → observed_at
}

impl ReplayCache {
    /// Create a new replay cache with given capacity and TTL.
    pub fn new(capacity: usize, ttl: Duration) -> Self {
        Self {
            inner: Arc::new(RwLock::new(Inner {
                capacity,
                ttl,
                map: IndexMap::with_capacity(capacity),
            })),
        }
    }

    /// Returns `true` if `job_id` has been seen within the TTL window.
    pub fn seen(&self, job_id: &str) -> bool {
        let inner = self.inner.read();
        match inner.map.get(job_id) {
            Some(observed_at) => observed_at.elapsed() < inner.ttl,
            None => false,
        }
    }

    /// Records `job_id` as seen. Evicts the oldest entry if at capacity.
    pub fn remember(&mut self, job_id: String) {
        let mut inner = self.inner.write();
        if inner.map.contains_key(&job_id) {
            // Update timestamp — move to end (most recent).
            inner.map.swap_remove(&job_id);
        } else if inner.map.len() >= inner.capacity {
            // Evict the oldest (first) entry.
            inner.map.shift_remove_index(0);
        }
        inner.map.insert(job_id, Instant::now());
    }

    /// Remove entries older than TTL. Call periodically from a background task.
    pub fn purge(&self) {
        let mut inner = self.inner.write();
        let ttl = inner.ttl;
        inner
            .map
            .retain(|_, observed_at| observed_at.elapsed() < ttl);
    }

    /// Returns the number of entries currently in the cache.
    pub fn len(&self) -> usize {
        self.inner.read().map.len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_seen_after_remember() {
        let mut cache = ReplayCache::new(10, Duration::from_secs(60));
        assert!(!cache.seen("job-1"));
        cache.remember("job-1".to_string());
        assert!(cache.seen("job-1"));
    }

    #[test]
    fn test_evict_oldest_at_capacity() {
        let mut cache = ReplayCache::new(3, Duration::from_secs(60));
        cache.remember("job-1".to_string());
        cache.remember("job-2".to_string());
        cache.remember("job-3".to_string());
        // Adding a 4th should evict job-1.
        cache.remember("job-4".to_string());
        assert!(!cache.seen("job-1"), "job-1 should have been evicted");
        assert!(cache.seen("job-4"));
    }

    #[test]
    fn test_replay_detection() {
        let mut cache = ReplayCache::new(10, Duration::from_secs(60));
        cache.remember("job-x".to_string());
        assert!(cache.seen("job-x"), "should be detected as replay");
    }
}
